src/tapdisk/datapath.py

```
@@ -59,6 +61,40 @@ def open(self, dbg, uri, persistent):
             raise xapi.storage.api.volume.Volume_does_not_exist(u.path)
         return None
 
+    def _metadata_dir(self, uri):
+        return TD_PROC_METADATA_DIR + "/" + uri
+
+    def save_tapdisk(self, dbg, uri, tap):
+        """ Record the tapdisk metadata for this URI in host-local storage """
+        dirname = self._metadata_dir(uri)
+        try:
+            os.makedirs(dirname, mode=0755)
+        except OSError as e:
+            if e.errno != 17: # 17 == EEXIST, which is harmless
+                raise e
+        with open(dirname + "/" + TD_PROC_METADATA_FILE, "w") as fd:
+            pickle.dump(tap.__dict__, fd)
+
+    def load_tapdisk(self, dbg, uri):
+        """ Recover the tapdisk metadata for this URI from host-local storage """
+        dirname = self._metadata_dir(uri)
+        if not(os.path.exists(dirname)):
+            # XXX throw a better exception
```

Shall we throw a better exception? ;)